### ISSUE TITLE
DNSCrypt causes big inefficiency

### ISSUE THREAD

**Issue by micielski** [erkexzcx/stoppropaganda#77]
I've just changed DNS to just Google and Cloudflare (primary, secondary) from DNSCrypt. I've been experiencing such a hit in 1gbps bandwidth on just only 1 normal and 1 dns worker that I could barely use the internet. Now everything works fine, being only the cpu as bottleneck but that's expected I guess :)
If anyone would confirm that DNSCrypt makes it very slow, that would be great!

**Comment by baylanger** [comment]
How are you configured when using DNSCrypt which OS?

Could it be when going over dnscrypt there's 0 caching and thus creates more traffic?!

**Comment by micielski** [comment]
I thought first time that changing DNS wouldn't help cause of caching. I have set it up properly, and changed the cache relatively big to what is mentioned by DNSCrypt's documentation.
in /var/cache there is a symbolic link to private/dnscrypt-proxy but it doesn't exist now.
I believe the reason is that I've disabled it and rebooted (privacy things I guess)
The OS is Arch Linux

**Comment by baylanger** [comment]
Regardless of DNScrypt or not, the OS should cache those as well at least the time defined by the TTL for each host.

If the TTL is 1m or below, caching longer time requires to "overwrite" the TTL.  I believe pi-hole allows that if not it's adguard.

My internal DNS goes like this : pihole -> adguard -> DoH Cloudflare.

**Comment by deputinizer** [comment]
With dnsmasq it works on raspberrypi very well.

You can set DNS servers to juice Putin a little.

Here's `/etc/myresolv.conf`
```
nameserver 194.54.14.186
nameserver 194.54.14.187
nameserver 84.252.147.118
```

**Comment by baylanger** [comment]
Good idea to use some of those!  Can't recall if by default libresolv automatically rotate between each or not.

**Comment by micielski** [comment]
Thanks for these advices on Raspberry Pi, and on DNS in general. I'll be having one Raspberry Pi in some days too, to set them t o dos too.

> Regardless of DNScrypt or not, the OS should cache those as well at least the time defined by the TTL for each host.
> If the TTL is 1m or below, caching longer time requires to "overwrite" the TTL. I believe pi-hole allows that if not it's adguard.
My TTL is high enough (min. 8400, max. 86400)

Speaking of things like that, setting timeout (not the DNS one) to 20 seconds with this change to fasthttp (I believe) unnecessarily greatly increases RAM usage (99% with 10GB RAM), so keep that in mind guys!

**Comment by erkexzcx** [comment]
> With dnsmasq it works on raspberrypi very well.
> 
> You can set DNS servers to juice Putin a little.
> 
> Here's /etc/myresolv.conf
> 
> nameserver 194.54.14.186
> nameserver 194.54.14.187
> nameserver 84.252.147.118

This should be in README.md recommendations :D

**Comment by baylanger** [comment]
@erkexzcx fyi ru DNS have changed again:

```
$ for ru_dns in $(dig ns ru|grep ^ru\.|awk '{print $NF}'); do host $ru_dns|grep "has address"|grep -E -o "[^^][0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}.[0-9]{1,3}"|sed -e 's/ //g'; done
193.232.128.6
194.85.252.62
194.190.124.17
193.232.142.17
193.232.156.17
```

Haven't look how everything gets built (docker, binary) ... can we automate this?  For moment I can daily check on those.

Do we need something in the app that can check the local /etc/resolv.conf in case end-user need to update it?!  It's tricky to handle this.

**Comment by baylanger** [comment]
And should we add Belarus NS has well 🤔

**Comment by baylanger** [comment]
Amazon is hosting these Belarus DNS 😠 

```
185.98.83.4
184.72.17.94
54.180.35.203
```

What can we do about this?
